libs/vm/include/vm/state.hpp

```diff
@@ -285,28 +291,40 @@ inline Ptr<IState> IState::Construct(VM *vm, TypeId type_id, Args &&... args)
   }  // switch
 }
 
-inline Ptr<IState> IState::Constructor(VM *vm, TypeId type_id, Ptr<String> const &name,
+template <typename... Args>
+inline Ptr<IState> IState::Construct(VM *vm, TypeId type_id, Args &&... args)
+{
+  TypeInfo const &type_info     = vm->GetTypeInfo(type_id);
+  TypeId const    value_type_id = type_info.parameter_type_ids[0];
+  return ConstructIntrinsic(vm, type_id, value_type_id, std::forward<Args>(args)...);
+}
+
+inline Ptr<IState> IState::Constructor(VM *vm, TypeId type_id, Ptr<Object> &&name,
                                        TemplateParameter const &value)
 {
   if (name)
   {
-    return Construct(vm, type_id, name, value);
+    return Construct(vm, type_id, std::move(name), value);
   }
 
-  vm->RuntimeError("Failed to construct State: name is null");
+  vm->RuntimeError("Failed to construct State: address is null");
   return nullptr;
 }
 
-inline Ptr<IState> IState::Constructor(VM *vm, TypeId type_id, Ptr<Address> const &address,
-                                       TemplateParameter const &value)
+inline std::string IState::NameToString(VM *vm, Ptr<Object> &&name)
 {
-  if (address)
+  switch (name->getTypeId())
   {
-    return Construct(vm, type_id, address->AsString(), value);
+  case TypeIds::String:
+    return dynamic_cast<String const &>(*name).str;
+  case TypeIds::Address:
+    return dynamic_cast<Address &>(*name).AsString()->str;
```

Comment:
why not const ref? Also, what happens if this throws?

Reply:
Because `AsString()` method is not declared as `const`.
I wanted to change that (a other similar things) in another commit, but we can that now.

Reply:
Regarding the second comment about _what happens if it throws_: We can ask exactly the same question about almost every line in the VM code - we are calling methods to dependencies - for example serialisation, IO observer, constructors, etc. ..., where almost evry such a call can potentially throw. Even `vm->RuntimeError(...)` can potentially throw.

VM has protection mechanism - the top level try-catch block where the otherwise uncaught exception will be caught and properly handled.